anywidget/_file_contents.py

```diff
@@ -0,0 +1,84 @@
+from __future__ import annotations
+from collections import deque
+
+import pathlib
+import threading
+from typing import Iterator
+
+from psygnal import Signal
+
+
+class FileContents:
+    """Object that watches for file changes and emits a signal when it changes.
+
+    Calling `str(obj)` on this object will always return the current contents of the
+    file (as long as the thread is running).
+
+    Parameters
+    ----------
+    path : str | pathlib.Path
+        The file to read and watch for content changes
+    start_thread : bool, optional
+        Whether to start watching for changes in a separate thread (default: `True`)
+    """
+
+    changed = Signal(str)
+    deleted = Signal()
+
+    def __init__(self, path: str | pathlib.Path, start_thread: bool = True):
+        self._path = pathlib.Path(path).expanduser().absolute()
+        if not self._path.is_file():
+            raise ValueError("File does not exist: {self._path}")
+        self._contents: str | None = None  # cached contents, cleared on change
+        self._stop_event = threading.Event()
+        self._background_thread: threading.Thread | None = None
+        if start_thread:
+            self.watch_in_thread()
+
+    def watch_in_thread(self) -> None:
+        """Watch for file changes (and emitting signals) from a separate thread."""
+        if self._background_thread is not None:
+            return
+        self._stop_event.clear()
+        self._background_thread = threading.Thread(
+            target=lambda: deque(self.watch(), maxlen=0),
```

Comment:
tell me about this line.  if the lambda and deque are added to make it work (as opposed to just using `target=self.watch`) then we could also consider removing the yield from the `watch` method (which turns it into a generator function).

Reply:
maybe nevermind... I see that you used `obj.watch` in tests...  which *is* nice if you want to retain the non-threaded behavior

Reply:
That's exactly it. `target=self.watch` just has the thread execute `self.watch()` which immediately returns a generator, so the the thread would terminate without entering an indefinite loop.

I saw that `deque(generator, maxlen=0)` is an [efficient way to consume a iterable](https://stackoverflow.com/questions/47456631/simpler-way-to-run-a-generator-function-without-caring-about-items) so I added that to have the thread exhaust the generator. You're right that the we could remove the `yield` keyword (changing `watch` from a generator to a regular function) and remove the need for this. I'm fine either way.

Reply:
nah, I later I realized that you do use it in tests... so it's nice to keep the functionality. let's stick with it.

Reply:
Sounds good. I'll add a comment